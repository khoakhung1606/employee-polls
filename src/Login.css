.login-container {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  background-color: #f8f9fa;
}

.login-card {
  text-align: center;
  background: #fff;
  padding: 2rem;
  border-radius: 10px;
  box-shadow: 0px 0px 15px rgba(0, 0, 0, 0.2);
}

.login-image {
  width: 150px;
  margin-bottom: 20px;
}

h2 {
  margin-bottom: 20px;
  font-size: 24px;
}

.input-group {
  margin-bottom: 15px;
}

input[type="text"],
input[type="password"] {
  width: 100%;
  padding: 10px;
  margin-top: 5px;
  border: 1px solid #ced4da;
  border-radius: 5px;
}

button {
  width: 100%;
  padding: 10px;
  background-color: #007bff;
  color: #fff;
  border: none;
  border-radius: 5px;
  font-size: 16px;
  cursor: pointer;
}

button:hover {
  background-color: #0056b3;
}
.error-message {
  color: red; /* Màu chữ đỏ để hiển thị lỗi */
  background-color: #f8d7da; /* Màu nền nhạt cho thông báo lỗi */
  border: 1px solid #f5c6cb; /* Viền mỏng cho khung thông báo lỗi */
  padding: 10px; /* Khoảng cách bên trong */
  border-radius: 5px; /* Bo góc mềm mại */
  font-size: 14px; /* Cỡ chữ */
  margin-top: 10px; /* Khoảng cách phía trên */
}
