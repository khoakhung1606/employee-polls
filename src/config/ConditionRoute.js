import { Navigate } from "react-router-dom";
import { connect } from "react-redux";

const ConditionRoute = ({ children, isLogin }) => {
  const url = window.location.href.toString().split(window.location.host)[1];

  return isLogin ? (
    children
  ) : (
    <Navigate to={url === "/" ? "/login" : `/login?redirect=${url}`} />
  );
};

const mapStateToProps = (state) => ({
  isLogin: !!state.userReducer.user,
});

export default connect(mapStateToProps)(ConditionRoute);
