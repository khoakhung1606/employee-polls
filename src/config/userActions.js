import { _getUsers } from "../_DATA";

export const SET_USER = "SET_USER";
export const SET_IS_LOGOUT = "SET_IS_LOGOUT";
export const SET_LIST_USER = "SET_LIST_USER";
export const LOGOUT_QUESTION = "LOGOUT_QUESTION";
export const LOGIN_FAILURE = "LOGIN_FAILURE";

export const setUser = (user) => ({
  type: SET_USER,
  user,
});

export const setIsLogout = (isLogout) => ({
  type: SET_IS_LOGOUT,
  isLogout,
});
export const loginFailure = (error) => ({
  type: LOGIN_FAILURE,
  error,
});

export const setListUser = (listUser) => ({
  type: SET_LIST_USER,
  listUser,
});

export const logoutQuestion = () => ({
  type: LOGOUT_QUESTION,
});

export const login = (id, password) => (dispatch, getState) => {
  const listUser = getState().userReducer.listUser;
  const user = Object.values(listUser).find((user) => user.id === id);

  setTimeout(() => {
    if (user && user.password === password) {
      dispatch(setUser(user));
    } else {
      dispatch(loginFailure("Invalid username or password"));
    }
  }, 300);
};

export const logout = () => (dispatch) => {
  setTimeout(() => {
    dispatch(setIsLogout(true));
    dispatch(logoutQuestion());
    dispatch(setListUser(null));
    dispatch(setUser(null));
  }, 300);
  return true;
};

export const getListUser = () => async (dispatch) => {
  const users = await _getUsers();
  console.log(users);
  dispatch(setListUser(users));
};
