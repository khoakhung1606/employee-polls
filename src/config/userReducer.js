import {
  SET_LIST_USER,
  SET_IS_LOGOUT,
  SET_USER,
  LOGIN_FAILURE,
} from "./userActions";

const defaultUserState = {
  user: null,
  listUser: null,
  isLogout: false,
};

const userReducer = (state = defaultUserState, action) => {
  switch (action.type) {
    case SET_USER:
      return { ...state, user: action.user };
    case SET_IS_LOGOUT:
      return { ...state, isLogout: action.isLogout };
    case SET_LIST_USER:
      return { ...state, listUser: action.listUser };
    case LOGIN_FAILURE:
      return { ...state, error: action.error }; // Set error on failure
    default:
      return state;
  }
};

export default userReducer;
