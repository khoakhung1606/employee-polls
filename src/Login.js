import React, { useState } from "react";
import { connect } from "react-redux";
import { Navigate } from "react-router-dom";
import "./Login.css";
import { login } from "./config/userActions";

function Login(props) {
  const { isLogin, login, error } = props;
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  const handleLogin = (e) => {
    e.preventDefault();
    login(username, password);
  };

  if (isLogin) {
    const params = new URLSearchParams(window.location.search);
    const url = params.get("redirect");
    return <Navigate to={url ? url : "/"} />;
  }

  return (
    <div className="login-container">
      <div className="login-card">
        <h2>Employee Polls</h2>
        <img src="login.PNG" alt="Employee Polls" className="login-image" />
        <form onSubmit={handleLogin}>
          <div data-testid="err_msg">
            {error && <p className="error-message">{error}</p>}
          </div>
          <div className="input-group">
            <label data-testid="user_label" htmlFor="username">
              User
            </label>
            <input
              data-testid="user_input"
              type="text"
              id="username"
              name="username"
              value={username}
              onChange={(e) => setUsername(e.target.value)}
              placeholder="User"
            />
          </div>
          <div className="input-group">
            <label data-testid="password_label" htmlFor="password">
              Password
            </label>
            <input
              data-testid="password_input"
              type="password"
              id="password"
              name="password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              placeholder="Password"
            />
          </div>
          <button data-testid="btn_submit" type="submit">
            Submit
          </button>
        </form>
      </div>
    </div>
  );
}
const mapStateToProps = (state) => {
  return {
    isLogin: !!state.userReducer.user,
    listUser: state.userReducer.listUser,
    error: state.userReducer.error,
  };
};

const mapDispatchtoProps = (dispatch) => {
  return {
    login: (name, password) => {
      dispatch(login(name, password));
    },
  };
};

export default connect(mapStateToProps, mapDispatchtoProps)(Login);
