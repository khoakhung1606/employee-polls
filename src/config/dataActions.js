import { _getQuestions, _saveQuestion, _saveQuestionAnswer } from "../_DATA";
import { setListUser } from "./userActions";

export const SET_LIST_QUESTIONS = "SET_LIST_QUESTIONS";
export const SET_ANSWERED = "SET_ANSWERED";
export const SET_NEW_QUESTIONS = "SET_NEW_QUESTIONS";

export const setListQuestions = (listQuestion) => ({
  type: SET_LIST_QUESTIONS,
  listQuestion,
});

export const setAnswered = (answered) => ({
  type: SET_ANSWERED,
  answered,
});

export const setNewQuestions = (newQuestion) => ({
  type: SET_NEW_QUESTIONS,
  newQuestion,
});

export const getListQuestion = () => (dispatch) => {
  return _getQuestions().then((questions) => {
    dispatch(setListQuestions(questions));
  });
};

export const addQuestion =
  (optionOneText, optionTwoText) => (dispatch, getState) => {
    const { id: author } = getState().userReducer.user;

    return _saveQuestion({ optionOneText, optionTwoText, author }).then(
      (question) => {
        let listUser = getState().userReducer.listUser;
        let listQuestion = getState().questionReducer.listQuestion;
        const updatedListQuestions = {
          ...listQuestion,
          [question.id]: question,
        };
        const updatedListUsers = {
          ...listUser,
          [question.author]: {
            ...listUser[question.author],
            questions: [...listUser[question.author].questions, question.id],
          },
        };

        dispatch(setListQuestions(updatedListQuestions));
        dispatch(setNewQuestions(question));
        dispatch(setListUser(updatedListUsers));
      }
    );
  };

export const saveQuestionAnswer =
  (authedUser, qid, answer) => (dispatch, getState) => {
    return _saveQuestionAnswer({ authedUser, qid, answer }).then(() => {
      let listUser = getState().userReducer.listUser;
      let listQuestion = getState().questionReducer.listQuestion;
      const updatedListUsers = {
        ...listUser,
        [authedUser]: {
          ...listUser[authedUser],
          answers: {
            ...listUser[authedUser].answers,
            [qid]: answer,
          },
        },
      };
      const updatedListQuestions = {
        ...listQuestion,
        [qid]: {
          ...listQuestion[qid],
          [answer]: {
            ...listQuestion[qid][answer],
            votes: [...listQuestion[qid][answer].votes, authedUser],
          },
        },
      };

      dispatch(setListQuestions(updatedListQuestions));
      dispatch(setAnswered({ authedUser, qid, answer }));
      dispatch(setListUser(updatedListUsers));
    });
  };
