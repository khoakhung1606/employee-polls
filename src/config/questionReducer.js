import {
  SET_LIST_QUESTIONS,
  SET_ANSWERED,
  SET_NEW_QUESTIONS,
} from "./dataActions";

import { LOGOUT_QUESTION } from "./userActions";

const defaultQuestionState = {
  listQuestion: null,
  answered: null,
  newQuestion: null,
};

const questionReducer = (state = defaultQuestionState, action) => {
  switch (action.type) {
    case SET_LIST_QUESTIONS:
      return { ...state, listQuestion: action.listQuestion };
    case SET_ANSWERED:
      return { ...state, answered: action.answered };
    case SET_NEW_QUESTIONS:
      return { ...state, newQuestion: action.newQuestion };
    case LOGOUT_QUESTION:
      return defaultQuestionState;
    default:
      return state;
  }
};

export default questionReducer;
