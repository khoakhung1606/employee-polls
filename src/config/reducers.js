import { combineReducers } from "redux";
import userReducer from "./userReducer";
import questionReducer from "./questionReducer";

export default combineReducers({
  userReducer,
  questionReducer,
});
