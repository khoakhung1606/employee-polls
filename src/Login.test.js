import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "./config/store";
import { BrowserRouter as Router } from "react-router-dom";
import Login from "./Login";
import { loginFailure, setUser, setListUser } from "./config/userActions";

describe("Login Component", () => {
  store.dispatch(
    setListUser({
      sarahedo: {
        id: "sarahedo",
        password: "password123",
        name: "Sarah Edo",
        avatarURL: null,
        answers: {
          "8xf0y6ziyjabvozdd253nd": "optionOne",
          "6ni6ok3ym7mf1p33lnez": "optionOne",
          am8ehyc8byjqgar0jgpub9: "optionTwo",
          loxhs1bqm25b708cmbf3g: "optionTwo",
        },
        questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
      },
      tylermcginnis: {
        id: "tylermcginnis",
        password: "abc321",
        name: "Tyler McGinnis",
        avatarURL: null,
        answers: {
          vthrdm985a262al8qx3do: "optionOne",
          xj352vofupe1dqz9emx13r: "optionTwo",
        },
        questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
      },
      mtsamis: {
        id: "mtsamis",
        password: "xyz123",
        name: "Mike Tsamis",
        avatarURL: null,
        answers: {
          xj352vofupe1dqz9emx13r: "optionOne",
          vthrdm985a262al8qx3do: "optionTwo",
          "6ni6ok3ym7mf1p33lnez": "optionOne",
        },
        questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
      },
    })
  );

  test("renders login form", () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    const userLabel = component.getByTestId("user_label");
    const passwordLabel = component.getByTestId("password_label");
    const btnSubmit = component.getByTestId("btn_submit");
    const errMsg = component.getByTestId("err_msg");

    expect(userLabel.textContent).toBe("User");
    expect(passwordLabel.textContent).toBe("Password");
    expect(btnSubmit.textContent).toBe("Submit");
    expect(errMsg.textContent).toBe("");
    expect(component).toMatchSnapshot();
  });

  test("displays error message when login fails", async () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const btnSubmit = component.getByTestId("btn_submit");
    fireEvent.submit(btnSubmit);

    await waitFor(() => {
      const errMsg = component.getByTestId("err_msg");
      expect(errMsg.textContent).toBe("Invalid username or password");
    });
    expect(component).toMatchSnapshot();
  });

  test("calls login on form submit and login success", async () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const userInput = component.getByTestId("user_input");
    const passwordInput = component.getByTestId("password_input");
    const btnSubmit = component.getByTestId("btn_submit");

    fireEvent.change(userInput, { target: { value: "mtsamis" } });
    fireEvent.change(passwordInput, { target: { value: "xyz123" } });
    expect(userInput.value).toBe("mtsamis");
    expect(passwordInput.value).toBe("xyz123");

    fireEvent.submit(btnSubmit);

    await waitFor(() => {
      expect(store.getState().userReducer.user).toEqual(
        setUser({
          id: "mtsamis",
          password: "xyz123",
          name: "Mike Tsamis",
          avatarURL: null,
          answers: {
            xj352vofupe1dqz9emx13r: "optionOne",
            vthrdm985a262al8qx3do: "optionTwo",
            "6ni6ok3ym7mf1p33lnez": "optionOne",
          },
          questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
        }).user
      );
    });
    expect(userInput).not.toBeInTheDocument();
    expect(passwordInput).not.toBeInTheDocument();
    expect(btnSubmit).not.toBeInTheDocument();
    expect(component).toMatchSnapshot();
  });
});
