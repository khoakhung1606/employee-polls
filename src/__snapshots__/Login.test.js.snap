// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component calls login on form submit and login success 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login Component displays error message when login fails 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <div
          class="login-card"
        >
          <h2>
            Employee Polls
          </h2>
          <img
            alt="Employee Polls"
            class="login-image"
            src="login.PNG"
          />
          <form>
            <div
              data-testid="err_msg"
            >
              <p
                class="error-message"
              >
                Invalid username or password
              </p>
            </div>
            <div
              class="input-group"
            >
              <label
                data-testid="user_label"
                for="username"
              >
                User
              </label>
              <input
                data-testid="user_input"
                id="username"
                name="username"
                placeholder="User"
                type="text"
                value=""
              />
            </div>
            <div
              class="input-group"
            >
              <label
                data-testid="password_label"
                for="password"
              >
                Password
              </label>
              <input
                data-testid="password_input"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <button
              data-testid="btn_submit"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <div
        class="login-card"
      >
        <h2>
          Employee Polls
        </h2>
        <img
          alt="Employee Polls"
          class="login-image"
          src="login.PNG"
        />
        <form>
          <div
            data-testid="err_msg"
          >
            <p
              class="error-message"
            >
              Invalid username or password
            </p>
          </div>
          <div
            class="input-group"
          >
            <label
              data-testid="user_label"
              for="username"
            >
              User
            </label>
            <input
              data-testid="user_input"
              id="username"
              name="username"
              placeholder="User"
              type="text"
              value=""
            />
          </div>
          <div
            class="input-group"
          >
            <label
              data-testid="password_label"
              for="password"
            >
              Password
            </label>
            <input
              data-testid="password_input"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            data-testid="btn_submit"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login Component renders login form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <div
          class="login-card"
        >
          <h2>
            Employee Polls
          </h2>
          <img
            alt="Employee Polls"
            class="login-image"
            src="login.PNG"
          />
          <form>
            <div
              data-testid="err_msg"
            />
            <div
              class="input-group"
            >
              <label
                data-testid="user_label"
                for="username"
              >
                User
              </label>
              <input
                data-testid="user_input"
                id="username"
                name="username"
                placeholder="User"
                type="text"
                value=""
              />
            </div>
            <div
              class="input-group"
            >
              <label
                data-testid="password_label"
                for="password"
              >
                Password
              </label>
              <input
                data-testid="password_input"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <button
              data-testid="btn_submit"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <div
        class="login-card"
      >
        <h2>
          Employee Polls
        </h2>
        <img
          alt="Employee Polls"
          class="login-image"
          src="login.PNG"
        />
        <form>
          <div
            data-testid="err_msg"
          />
          <div
            class="input-group"
          >
            <label
              data-testid="user_label"
              for="username"
            >
              User
            </label>
            <input
              data-testid="user_input"
              id="username"
              name="username"
              placeholder="User"
              type="text"
              value=""
            />
          </div>
          <div
            class="input-group"
          >
            <label
              data-testid="password_label"
              for="password"
            >
              Password
            </label>
            <input
              data-testid="password_input"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            data-testid="btn_submit"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
